R/ptype-abbr-full.R

```diff
@@ -32,15 +32,25 @@
 #' cat(vec_ptype_abbr(1:10))
 vec_ptype_full <- function(x, ...) {
   check_dots_empty0(...)
-  UseMethod("vec_ptype_full")
+
+  # Data frames dispatch to the `data.frame` method by default to get
+  # the inner types format
+  method <- s3_method_specific(
+    x,
+    "vec_ptype_full",
+    ns = "vctrs",
+    df_default = TRUE
+  )
+  method(x, ...)
 }
 
 #' @export
 #' @rdname vec_ptype_full
 vec_ptype_abbr <- function(x, ..., prefix_named = FALSE, suffix_shape = TRUE) {
   check_dots_empty0(...)
 
-  abbr <- vec_ptype_abbr_dispatch(x)
+  method <- s3_method_specific(x, "vec_ptype_abbr", ns = "vctrs")
+  abbr <- method(x, ...)
```

Comment:
I don't think this is quite right yet for data frames. The `vec_ptype_abbr()` default will call `vec_ptype_full()` and abbreviate it, which can result in some ugly results for things like data tables, which don't have a `vec_ptype_abbr()` method (unlike tibbles, provided by the tibble package).

``` r
library(vctrs)

df <- tibble::tibble(x = 1, y = 2, z = 3)

# tibble:::vec_ptype_full.tbl_df does not exist.
# uses the data frame fallback in `vec_ptype_full()` because `df_default = T`
vec_ptype_full(df)
#> [1] "tbl_df<\n  x: double\n  y: double\n  z: double\n>"

# tibble:::vec_ptype_abbr.tbl_df exists so this is nice
vec_ptype_abbr(df)
#> [1] "tibble[,3]"

library(data.table)

dt <- data.table(x = 1, y = 2, z = 3)

# again uses the data frame fallback
vec_ptype_full(dt)
#> [1] "data.table<\n  x: double\n  y: double\n  z: double\n>"

# there is no data table method for vec_ptype_abbr, so it falls through to
# vec_ptype_abbr.default, which abbreviates the vec_ptype_full result in
# an ugly way
vec_ptype_abbr(dt)
#> [1] "dxdydzd>[,3]"
```

<sup>Created on 2022-06-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Would be great to see a data frame test or two added as well once this gets fixed

Reply:
Good catch. But isn't this a problem in general with the `is.object` branch of this default "abbr" method that first calls "full" and then abbreviate it?

Reply:
I don't see any broken test when I change the method to:

```r
vec_ptype_abbr.default <- function(x, ...) {
  if (is.object(x)) {
    type <- class(x)[[1]]
  } else if (is_vector(x)) {
    type <- vec_ptype_abbr_bare(x, ...)
  } else {
    abort("Not a vector.")
  }
  unname(abbreviate(type, 8))
}
```

We should probably do that.